Add CI and data integrity tests for SMILES

cthoyt opened the pull request:
Closes #1

This PR does the following:

1. Adds a `tox.ini` file for making testing and linting reproducible from any environment
2. Adds tests for all SMILES being non-null and parsable. There are a handful of cases where this isn't true that I suggest we fix before accepting this PR
3. Adds code quality tests to make sure that the python code that goes in here stays readable and maintainable
   - `flake8` checks code quality
   - `black` automatically reformats code to look nice
   - `mypy` makes sure that code is self consistent using optional type hints
4. Adds GitHub actions configuration to make sure that these tests and code quality are run on each commit. In the future, you should not accept any contributions that break the build. I've found in my own curation projects than an excellent workflow is to make a new branch, write a test that corresponds to the curation I will do, then do the curation.

## Required curation

There is one lines which do not contain a SMILES string at all:
- 21446 (3,8-Dimethylceroxen)

I've started chipping away at the remaining issues by doing some of my own curation, but I think I'll need a bt of help for the last few (I have a background in med/synthetic chem, but they're still tricky). The CI is run at [![Tests](https://github.com/cthoyt/GuthrieSolv/actions/workflows/tests.yml/badge.svg)](https://github.com/cthoyt/GuthrieSolv/actions/workflows/tests.yml) every time I make a commit. If you accept that my PR can run GitHub actions on this repo, then you can follow the links directly at the bottom of the PR.

cthoyt committed:
Add CI and data integrity tests for SMILES

Closes #1

cthoyt committed:
Switch to unittest

cthoyt committed:
Update .gitignore

cthoyt committed:
Add first curations

cthoyt committed:
Improve tests feedback and more curations

cthoyt committed:
Update guthrie_database.csv

cthoyt committed:
Update test_data.py

cthoyt committed:
Update guthrie_database.csv

cthoyt committed:
Update guthrie_database.csv

davidlmobley reviewed (approved):
Thanks! This looks like a great start; I've not used the Tox testing you're bringing in here but I'm going to trust it works. Do you want to propose an update to the README.md which would explain how to use it/that it's here? You might also propose an update there which adds a changelog and a contributors list and adds yourself to the latter. :) 

Also let me know if it'd help if you have write access to the repo.

cthoyt committed:
Could not find in paper

I was able to track down the paper https://www.jstage.jst.go.jp/article/nikkashi1972/1977/2/1977_2_264/_article/-char/en but not figure out which compound this was referring to

cthoyt committed:
Could not find secondary reference

The paper had a reference to another book I could not get my hands on, but this nomenclature is totally nonsensical since ceroxen just has 3 carbons in it.

cthoyt committed:
Finish cleanup

cthoyt committed:
Update test_data.py

cthoyt commented:
I finished curation and updated the readme. This should now be ready to merge. I’ll make some more structural changes to the readme for contribution information in a future PR. I think it would be great if you grant me write access!

davidlmobley commented:
Thank you!
